pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -291,57 +292,50 @@ func (og *operationGenerator) GenerateBulkVolumeVerifyFunc(
 
 func (og *operationGenerator) GenerateAttachVolumeFunc(
 	volumeToAttach VolumeToAttach,
-	actualStateOfWorld ActualStateOfWorldAttacherUpdater) (volumetypes.GeneratedOperations, error) {
-	var err error
-	var attachableVolumePlugin volume.AttachableVolumePlugin
-
+	actualStateOfWorld ActualStateOfWorldAttacherUpdater) volumetypes.GeneratedOperations {
 	// Get attacher plugin
-	eventRecorderFunc := func(err *error) {
-		if *err != nil {
-			for _, pod := range volumeToAttach.ScheduledPods {
-				og.recorder.Eventf(pod, v1.EventTypeWarning, kevents.FailedAttachVolume, (*err).Error())
-			}
-		}
-	}
-
-	originalSpec := volumeToAttach.VolumeSpec
-	nu, err := nodeUsingCSIPlugin(og, volumeToAttach.VolumeSpec, volumeToAttach.NodeName)
-	if err != nil {
-		eventRecorderFunc(&err)
-		return volumetypes.GeneratedOperations{}, volumeToAttach.GenerateErrorDetailed("AttachVolume.NodeUsingCSIPlugin failed", err)
+	attachableVolumePluginName := unknownAttachableVolumePlugin
+	attachableVolumePlugin, err :=
+		og.volumePluginMgr.FindAttachablePluginBySpec(volumeToAttach.VolumeSpec)
+	if err == nil && attachableVolumePlugin != nil {
+		attachableVolumePluginName = attachableVolumePlugin.GetPluginName()
 	}
 
-	// useCSIPlugin will check both CSIMigration and the plugin specific feature gate
-	if useCSIPlugin(og.volumePluginMgr, volumeToAttach.VolumeSpec) && nu {
-		// The volume represented by this spec is CSI and thus should be migrated
-		attachableVolumePlugin, err = og.volumePluginMgr.FindAttachablePluginByName(csi.CSIPluginName)
-		if err != nil || attachableVolumePlugin == nil {
-			eventRecorderFunc(&err)
-			return volumetypes.GeneratedOperations{}, volumeToAttach.GenerateErrorDetailed("AttachVolume.FindAttachablePluginByName failed", err)
-		}
-
-		csiSpec, err := translateSpec(volumeToAttach.VolumeSpec)
+	attachVolumeFunc := func() (error, error) {
+		var attachableVolumePlugin volume.AttachableVolumePlugin
+		originalSpec := volumeToAttach.VolumeSpec
+		nu, err := nodeUsingCSIPlugin(og, volumeToAttach.VolumeSpec, volumeToAttach.NodeName)
 		if err != nil {
-			return volumetypes.GeneratedOperations{}, volumeToAttach.GenerateErrorDetailed("AttachVolume.TranslateSpec failed", err)
+			return volumeToAttach.GenerateError("AttachVolume.NodeUsingCSIPlugin failed", err)
 		}
 
-		volumeToAttach.VolumeSpec = csiSpec
-	} else {
-		attachableVolumePlugin, err =
-			og.volumePluginMgr.FindAttachablePluginBySpec(volumeToAttach.VolumeSpec)
-		if err != nil || attachableVolumePlugin == nil {
-			eventRecorderFunc(&err)
-			return volumetypes.GeneratedOperations{}, volumeToAttach.GenerateErrorDetailed("AttachVolume.FindAttachablePluginBySpec failed", err)
+		// useCSIPlugin will check both CSIMigration and the plugin specific feature gate
+		if useCSIPlugin(og.volumePluginMgr, volumeToAttach.VolumeSpec) && nu {
```

Comment:
I'm not sure about this if/else logic can be simplified like [this](https://gist.github.com/mucahitkurt/d6934c9a474ba3e862a1740c83cf23b9)

Reply:
Seems reasonable to me. But I will let one of the CSI migration folks(@ddebroy @davidz627 @leakingtapan) verify.

Reply:
SGTM too.

Reply:
So, I'll simplify this

Reply:
I was just speaking with @leakingtapan and he pointed out that @mkimuram raised some concerns around potential issues with handling reconstruction failures at https://github.com/kubernetes/kubernetes/pull/74844#discussion_r266272782 when he was thinking of switching from `FindAttachablePluginByName ` to `FindAttachablePluginBySpec`. @mkimuram would there be a similar concern here too with the proposed [simplification](https://gist.github.com/mucahitkurt/d6934c9a474ba3e862a1740c83cf23b9)

Reply:
Thanks, @leakingtapan pointed out that discussion to me as an answer to my question in sig storage mailing list, when I checked the `GenerateAttachVolumeFunc` usage paths, I couldn't see the similar usage path(not setting the `volumeToAttach.VolumeSpec`).

I see the similar usage of my proposed simplification in [`GenerateMountVolumeFunc`](https://github.com/kubernetes/kubernetes/blob/990ee3c09c0104cc1045b343040fe76082862d73/pkg/volume/util/operationexecutor/operation_generator.go#L529). But usage scenarios can be different.

Reply:
@vladimirvivien opened an [issue](https://github.com/kubernetes/kubernetes/issues/75838) about the usage of find plugin by spec vs find plugin by name, and he also pointed out this function.

@vladimirvivien what do you think about my simplification? Thanks!

Reply:
@ddebroy, @leakingtapan I'll rollback this simplification to make this PR go forward.